src/components/settings/PersonalizedPushSettings.tsx

```diff
@@ -0,0 +1,105 @@
+import React, { useContext, useEffect, useMemo, useState } from 'react';
+import { useQuery } from 'react-apollo';
+import { FlatList, StyleSheet } from 'react-native';
+
+import { SettingsContext } from '../../SettingsProvider';
+import { normalize, texts } from '../../config';
+import { getInAppPermission } from '../../pushNotifications';
+import { getQuery, QUERY_TYPES } from '../../queries';
+import { onActivatePushNotifications, onDeactivatePushNotifications } from '../../screens';
+import { LoadingSpinner } from '../LoadingSpinner';
+import { SafeAreaViewFlex } from '../SafeAreaViewFlex';
+import { SettingsToggle } from '../SettingsToggle';
+
+type Category = { id: string; iconName?: string | null; name: string };
+
+export const PersonalizedPushSettings = () => {
+  const { globalSettings } = useContext(SettingsContext);
+  const { settings = {} } = globalSettings;
+  const { personalizedPush = {}, settingsScreenListItemTitles = {} } = settings;
+  const queryVariables = personalizedPush.variables || {
+    tagList: ['news_item']
+  };
+
+  const [permission, setPermission] = useState<boolean>(false);
+
+  const { data, loading } = useQuery(getQuery(QUERY_TYPES.CATEGORIES_FILTER), {
+    variables: queryVariables,
+    skip: !personalizedPush?.categories
+  });
+
+  useEffect(() => {
+    (async () => {
+      const inAppPermission = await getInAppPermission();
+      setPermission(!!inAppPermission);
+    })();
+  }, []);
+
+  const onActivate = (revert: () => void) => {
+    setPermission(true);
+    onActivatePushNotifications(() => {
+      setPermission(false);
+      revert?.();
+    });
+  };
+
+  const onDeactivate = (revert: () => void) => {
+    setPermission(false);
+    onDeactivatePushNotifications(() => {
+      setPermission(true);
+      revert?.();
+    });
+  };
```

Comment:
can you explain this to me please in a few own words? when and what is `revert` method triggering?

Reply:
I had to use this method here because I needed to update the state in this file to control the category push switchers according to the value of the main push switcher. After updating the state, I send the revert function I need. This allows the token to be registered or deleted on the server.